### PR TITLE
RES: Support impl for type parameter `impl<T> Trait for T`

### LINKED ISSUE #1735: Support impl for type parameter `impl<T> Trait for T`
Currently we don't look for impls like `impl<T> Trait for T`, so this test fails
```rust
trait Trait {
    fn foo(&self) {}
}    //X
impl<T> Trait for T {}
struct X;
fn main() {
    X.foo();
}   //^
```
It is pretty easy to implement. [Here](https://github.com/intellij-rust/intellij-rust/blob/9603f049d96b1ec515830b1db03737db14e3af1a/src/main/kotlin/org/rust/lang/core/resolve/indexes/RsImplIndex.kt#L48) we indexing the `impl` by the type. We should check if type is a type parameter of this impl (go to parent psi, find type parameters declaration and check the type name is the same), then if it is a type parameter, save it to index with a specific key that is invalid identifier (like #). Then while finding impls retreive index twice: with actual type name and with our specific key.

---
You can find more info about how to start contributing to Intellij-Rust in [CONTRIBUTING.md](https://github.com/intellij-rust/intellij-rust/blob/master/CONTRIBUTING.md), and some info about our architecure in [ARCHITECTURE.md](https://github.com/intellij-rust/intellij-rust/blob/master/ARCHITECTURE.md). If you have any questions about this issue, you can ask them right here or in our [gitter](https://gitter.im/intellij-rust/intellij-rust). If you decided to grab this issue, please leave a comment here and put yourself in the [worklist](https://github.com/intellij-rust/intellij-rust/wiki/Worklist).

### PR DESCRIPTION
**Pull request by Undin**
Closes #1735

### CODE REVIEW

**Comment by Undin** [comment]
This implementation checks if type parameter is unbounded because otherwise it leads to stack overflow. Solution of this problem may require a lot of changes in code and I think it should be fixed while supporting trait bounds in separate issue.

File: src/main/kotlin/org/rust/lang/core/resolve/indexes/RsImplIndex.kt [hunk 1]
```diff
@@ -33,19 +38,36 @@ class RsImplIndex : AbstractStubIndex<TyFingerprint, RsImplItem>() {
...
+            val impl = stub.psi
+            val typeRef = impl.typeReference ?: return
+            val type = typeRef.typeElement
+
+            val key = if (type is RsBaseType) {
+                val typeParam = impl.typeParameters.find { it.name == type.name }
+                // At this moment we support only free trait impls without bounds
+                // to avoid stack overflow while resolving
+                if (typeParam != null && typeParam.bounds.isEmpty()) {
+                    TyFingerprint.TYPE_PARAMETER_FINGERPRINT
```
Comment (vlad20012) [inline src/main/kotlin/org/rust/lang/core/resolve/indexes/RsImplIndex.kt]: If type params is not empty we should just ignore the impl (exclude from index)
Also, what about whew clauses?

**Comment by vlad20012** [comment]
Ok, I understood why SO here. So... this will be fixed when I finish replacing `Ty.unifyWith` with `FulfillmentContext`

File: src/main/kotlin/org/rust/lang/core/resolve/indexes/RsImplIndex.kt (hunk 1 above)
Comment (Undin) [reply src/main/kotlin/org/rust/lang/core/resolve/indexes/RsImplIndex.kt]: `RsTypeParameter.bounds` is extension property (see [here](https://github.com/intellij-rust/intellij-rust/blob/4c7c1e331e49b84aa262fcc7471cef72605a3920/src/main/kotlin/org/rust/lang/core/psi/ext/RsTypeParameter.kt#L17-L25])) which collect all bounds including where clause

Comment (vlad20012) [reply src/main/kotlin/org/rust/lang/core/resolve/indexes/RsImplIndex.kt]: Oh, OK

**Comment by vlad20012** [comment]
bors r+
Many thanks!

**Comment by vlad20012** [comment]
Oh, we forgot to bump stub/index version...

**Comment by Undin** [comment]
@vlad20012 Let's fix it!

**Comment by Undin** [comment]
Not very successful PR :(

**Pull request merged**